solana/restaking-v2/programs/restaking-v2/src/lib.rs

```diff
@@ -0,0 +1,436 @@
+pub const RECEIPT_SEED: &[u8] = b"receipt";
+
+pub const RECEIPT_TOKEN_DECIMALS: u8 = 9;
+
+#[cfg(test)]
+mod tests;
+
+#[program]
+pub mod restaking_v2 {
+    use anchor_spl::token::{Burn, MintTo, Transfer};
+
+    use super::*;
+
+    pub fn initialize(
+        ctx: Context<Initialize>,
+        whitelisted_tokens: Vec<Pubkey>,
+        initial_validators: Vec<Pubkey>,
+        guest_chain_program_id: Pubkey,
+    ) -> Result<()> {
+        msg!("Initializng Restaking program");
+
+        let common_state = &mut ctx.accounts.common_state;
+
+        common_state.admin = ctx.accounts.admin.key();
+        common_state.whitelisted_tokens = whitelisted_tokens;
+        common_state.validators = initial_validators;
+        common_state.guest_chain_program_id = guest_chain_program_id;
+
+        Ok(())
+    }
+
+    /// Deposit tokens in the escrow and mint receipt tokens to the staker while updating the
+    /// stake for the validators on the guest chain.
+    ///
+    /// Fails if
+    /// - token to be staked is not whitelisted
+    /// - staker does not have enough tokens
+    /// - accounts needed to call guest chain program are missing
+    pub fn deposit(ctx: Context<Deposit>, amount: u64) -> Result<()> {
+        let common_state = &mut ctx.accounts.common_state;
+
+        let stake_token_mint = &ctx.accounts.token_mint.key();
+
+        if common_state
+            .whitelisted_tokens
+            .iter()
+            .find(|&x| x == stake_token_mint)
+            .is_none()
+        {
+            return Err(error!(ErrorCodes::InvalidTokenMint));
+        }
+
+        if ctx.accounts.staker_token_account.amount < amount {
+            return Err(error!(ErrorCodes::NotEnoughTokensToStake));
+        }
+
+        let bump = ctx.bumps.common_state;
+        let seeds = [COMMON_SEED, core::slice::from_ref(&bump)];
+        let seeds = seeds.as_ref();
+        let seeds = core::slice::from_ref(&seeds);
+
+        let transfer_ix = Transfer {
+            from: ctx.accounts.staker_token_account.to_account_info(),
+            to: ctx.accounts.escrow_token_account.to_account_info(),
+            authority: ctx.accounts.staker.to_account_info(),
+        };
+
+        let cpi_ctx = CpiContext::new(ctx.accounts.token_program.to_account_info(), transfer_ix);
+
+        anchor_spl::token::transfer(cpi_ctx, amount)?;
+
+        let mint_to_ix = MintTo {
+            mint: ctx.accounts.receipt_token_mint.to_account_info(),
+            to: ctx.accounts.staker_receipt_token_account.to_account_info(),
+            authority: common_state.to_account_info(),
+        };
+
+        let cpi_ctx = CpiContext::new_with_signer(
+            ctx.accounts.token_program.to_account_info(),
+            mint_to_ix,
+            seeds,
+        );
+
+        anchor_spl::token::mint_to(cpi_ctx, amount)?;
+
+        // Call guest chain program to update the stake equally
+        let stake_per_validator = amount / common_state.validators.len() as u64;
```

Comment:
amount may be indivisible by the number of validators. We may want to add a reminder to a random validator.

Reply:
i think that makes sense.

Reply:
i think we should add the remainder to a particular validator because during withdrawal we might be withdrawing the remainder from a different validator.

Reply:
not that the remainder value would be high, but i think it can cause underflow during withdrawal. The other option might be that if it causes underflow, we remove it from a different validator but then we need to deserialize the chain data on restaking layer which i dont think is really worth it.